z_test_implementation.py

```diff
@@ -1,63 +1,65 @@
 import time
 
 
-def node_5s()->float:
+def node_5s() -> float:
     print("5s executed")
     start = time.time()
     time.sleep(5)
     return time.time() - start
 
-def node_5s_error()->float:
+
+def at_1_to_previous(node_5s: float) -> float:
```

Comment:
Added to see how two connecting nodes would respond and ran into a different issue:

```
Captured execution run. Results can be found at http://localhost:8241/dashboard/project/1/runs/101

(execute_lifecycle_for_node pid=631084) 5s executed
                                    at_1_to_previous
0  ObjectRef(c8ef45ccd0112571ffffffffffffffffffff...
```

Reply:
Hmm, the problem is it's getting passed an object ref rather than a function? Is this reflected in the UI? Or just the result?

Reply:
Here's the weird part that I am troubleshooting at the moment. Asking for the ```node_5s``` displays correctly on the UI that it took 5s to run.

I then added this function into the DAG
```
def at_1_to_previous(node_5s: float) -> float:
    print("Value of 5s ", node_5s)
    print("1s executed")
    start = time.time()
    time.sleep(1)
    return node_5s   + (time.time()-start)
```

If I request ```at_1_to_previous``` I get an error message and the UI shows the run but no nodes are listed as executed or failed, it's just an empty waterfall and list.

```
2024-08-24 21:56:32,934 INFO worker.py:1781 -- Started a local Ray instance.

Capturing execution run. Results can be found at http://localhost:8241/dashboard/project/1/runs/111


Captured execution run. Results can be found at http://localhost:8241/dashboard/project/1/runs/111

(execute_lifecycle_for_node pid=661223) Value of 5s  ObjectRef(c8ef45ccd0112571ffffffffffffffffffffffff0100000001000000)
(execute_lifecycle_for_node pid=661223) 1s executed
-------------------------------------------------------------------
Oh no an error! Need help with Hamilton?
Join our slack and ask for help! https://join.slack.com/t/hamilton-opensource/shared_invite/zt-2niepkra8-DGKGf_tTYhXuJWBTXtIs4g
-------------------------------------------------------------------

Traceback (most recent call last):
  File "/home/jfrank/code/hamilton/z_test_implementation.py", line 46, in <module>
    result_ray = dr_ray.execute(
  File "/home/jfrank/code/hamilton/hamilton/driver.py", line 567, in execute
    raise e
  File "/home/jfrank/code/hamilton/hamilton/driver.py", line 560, in execute
    return self.adapter.call_lifecycle_method_sync("do_build_result", outputs=outputs)
  File "/home/jfrank/code/hamilton/hamilton/lifecycle/base.py", line 959, in call_lifecycle_method_sync
    return getattr(adapter, method_name)(**kwargs)
  File "/home/jfrank/code/hamilton/hamilton/plugins/h_ray.py", line 154, in do_build_result
    result = ray.get(remote_combine)  # this materializes the object locally
  File "/home/jfrank/miniconda3/envs/hamilton/lib/python3.10/site-packages/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/home/jfrank/miniconda3/envs/hamilton/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/home/jfrank/miniconda3/envs/hamilton/lib/python3.10/site-packages/ray/_private/worker.py", line 2659, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/home/jfrank/miniconda3/envs/hamilton/lib/python3.10/site-packages/ray/_private/worker.py", line 871, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(TypeError): ray::build_result() (pid=661223, ip=10.21.249.30)
  At least one of the input arguments for this task could not be computed:
ray.exceptions.RayTaskError: ray::execute_lifecycle_for_node() (pid=661223, ip=10.21.249.30)
  File "/home/jfrank/code/hamilton/hamilton/execution/graph_functions.py", line 243, in execute_lifecycle_for_node
    result = node_(**kwargs)
  File "/home/jfrank/code/hamilton/hamilton/node.py", line 249, in __call__
    return self.callable(*args, **kwargs)
  File "/home/jfrank/code/hamilton/z_test_implementation.py", line 16, in at_1_to_previous
    return node_5s   + (time.time()-start)
TypeError: unsupported operand type(s) for +: 'ray._raylet.ObjectRef' and 'float'
(execute_lifecycle_for_node pid=661223) 
(execute_lifecycle_for_node pid=661223) ********************************************************************************
(execute_lifecycle_for_node pid=661223) > at_1_to_previous [__main__.at_1_to_previous()] encountered an error          <
(execute_lifecycle_for_node pid=661223) > Node inputs:
(execute_lifecycle_for_node pid=661223) {'node_5s': 'ObjectRef(c8ef45ccd0112571ffffffffffffffffffffffff...'}
(execute_lifecycle_for_node pid=661223) ********************************************************************************
(execute_lifecycle_for_node pid=661223) Traceback (most recent call last):
(execute_lifecycle_for_node pid=661223)   File "/home/jfrank/code/hamilton/hamilton/execution/graph_functions.py", line 243, in execute_lifecycle_for_node
(execute_lifecycle_for_node pid=661223)     result = node_(**kwargs)
(execute_lifecycle_for_node pid=661223)   File "/home/jfrank/code/hamilton/hamilton/node.py", line 249, in __call__
(execute_lifecycle_for_node pid=661223)     return self.callable(*args, **kwargs)
(execute_lifecycle_for_node pid=661223)   File "/home/jfrank/code/hamilton/z_test_implementation.py", line 16, in at_1_to_previous
(execute_lifecycle_for_node pid=661223)     return node_5s   + (time.time()-start)
(execute_lifecycle_for_node pid=661223) TypeError: unsupported operand type(s) for +: 'ray._raylet.ObjectRef' and 'float'
(execute_lifecycle_for_node pid=661226) 5s executed
```
Lastly, if I change the return to
```
def at_1_to_previous(node_5s: float) -> float:
    print("Value of 5s ", node_5s)
    print("1s executed")
    start = time.time()
    time.sleep(1)
    return node_5s   #+ (time.time()-start)
```
The UI again doesn't show anything and I get
```
/home/jfrank/miniconda3/envs/hamilton/bin/python /home/jfrank/code/hamilton/z_test_implementation.py
2024-08-24 21:59:25,593 INFO worker.py:1781 -- Started a local Ray instance.

Capturing execution run. Results can be found at http://localhost:8241/dashboard/project/1/runs/113


Captured execution run. Results can be found at http://localhost:8241/dashboard/project/1/runs/113

(execute_lifecycle_for_node pid=662808) Value of 5s  ObjectRef(c8ef45ccd0112571ffffffffffffffffffffffff0100000001000000)
(execute_lifecycle_for_node pid=662808) 1s executed
                                    at_1_to_previous
0  ObjectRef(c8ef45ccd0112571ffffffffffffffffffff...
(execute_lifecycle_for_node pid=662810) 5s executed
                                    at_1_to_previous
0  ObjectRef(c8ef45ccd0112571ffffffffffffffffffff...
```

Reply:
so without diving in the details above this is my mental model for what should happen.

1. Hamilton is building a delayed execution graph with ray.remote functions. This means that as kwargs input to a function we will pass around ray object references.
2. this happens "driver" side -- i.e. functions that actually run on ray have the object references resolved properly by ray, and are none the wiser.
3. so "driver side" the result from the node will include a ray object reference.
4. the builder resolves all of these to pull the results locally to the driver side.

Does that make sense? I will look more closely tonight.